apps/rush-lib/src/logic/pnpm/PnpmfileShim.ts

```diff
@@ -1,53 +1,72 @@
+  // So we need to ensure the context format before we move on
+  if (typeof context !== 'object' || Array.isArray(context)) {
+    context = {
+      log: (message: string) => {},
+      originalContext: context
+    } as IPnpmfileContext;
+  }
+  if (!settings) {
+    // Initialize the settings from file
+    if (!context.pnpmfileShimSettings) {
+      context.pnpmfileShimSettings = require('./pnpmfileSettings.json');
+    }
+    settings = context.pnpmfileShimSettings!;
+  } else if (!context.pnpmfileShimSettings) {
+    // Reuse the already initialized settings
+    context.pnpmfileShimSettings = settings;
+  }
+  if (!clientPnpmfile && settings.clientPnpmfilePath) {
+    clientPnpmfile = require(settings.clientPnpmfilePath);
+  }
+  if (!semver && settings.semverPath) {
+    semver = require(settings.semverPath);
+  }
+  return context as IPnpmfileContext;
 }
 
-// Load in the generated settings file
-const pnpmfileSettings: IPnpmfileShimSettings = require('./pnpmfileSettings.json');
-// We will require semver from this path on disk, since this is the version of semver shipping with Rush
-const semver: typeof TSemver = require(pnpmfileSettings.semverPath);
-// Only require the client pnpmfile if requested
-const clientPnpmfile: IPnpmfile | undefined = pnpmfileSettings.useClientPnpmfile
-  ? require('./clientPnpmfile')
-  : undefined;
-
 // Set the preferred versions on the dependency map. If the version on the map is an allowedAlternativeVersion
 // then skip it. Otherwise, check to ensure that the common version is a subset of the specified version. If
 // it is, then replace the specified version with the preferredVersion
-function setPreferredVersions(dependencies?: { [dependencyName: string]: string }): void {
+function setPreferredVersions(dependencies: { [dependencyName: string]: string } | undefined): void {
   for (const name of Object.keys(dependencies || {})) {
-    if (pnpmfileSettings.allPreferredVersions.hasOwnProperty(name)) {
-      const preferredVersion: string = pnpmfileSettings.allPreferredVersions[name];
+    if (settings.allPreferredVersions?.hasOwnProperty(name)) {
+      const preferredVersion: string = settings.allPreferredVersions[name];
       const version: string = dependencies![name];
-      if (pnpmfileSettings.allowedAlternativeVersions.hasOwnProperty(name)) {
+      if (settings.allowedAlternativeVersions?.hasOwnProperty(name)) {
         const allowedAlternatives: ReadonlyArray<string> | undefined =
-          pnpmfileSettings.allowedAlternativeVersions[name];
+          settings.allowedAlternativeVersions[name];
         if (allowedAlternatives && allowedAlternatives.indexOf(version) > -1) {
           continue;
         }
       }
       let isValidRange: boolean = false;
       try {
-        isValidRange = !!semver.validRange(preferredVersion) && !!semver.validRange(version);
+        isValidRange = !!semver!.validRange(preferredVersion) && !!semver!.validRange(version);
```

Comment:
Might make sense to compute this by constructing the semver.Range objects, since they'll get used again by subset and that way we don't construct them twice.